pkg/issuer/acme/dns/execute/execute.go

```diff
@@ -0,0 +1,103 @@
+package execute
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+var workingDirectory = "/srv/execute_plugin_dir/"
+
+type DNSProvider struct {
+	pluginName string
+	binaryPath string
+	env        []string
+}
+
+func NewDNSProvider() (*DNSProvider, error) {
+	pluginName := os.Getenv("EXECUTE_PLUGIN_NAME")
+	env := envStrToSlice(os.Getenv("EXECUTE_ENV"))
+
+	return NewDNSProviderCredentials(pluginName, env)
+}
+
+func NewDNSProviderCredentials(pluginName string, env []string) (*DNSProvider, error) {
+	if pluginName == "" || len(env) == 0 {
+		return nil, fmt.Errorf("execute parameters missing")
+	}
+
+	return &DNSProvider{
+		pluginName: pluginName,
+		binaryPath: workingDirectory + pluginName,
+		env:        env,
+	}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS
+// propagation. Adjusting here to cope with spikes in propagation times.
+func (c *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return 120 * time.Second, 2 * time.Second
+}
+
+func (c *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, _ := util.DNS01Record(domain, keyAuth)
+	args := []string{"present", util.UnFqdn(fqdn), value}
```

Comment:
Slightly nervous about taking user-inputted data and executing it on the CLI.

Perhaps serialising this data to JSON or similar, and passing it over STDIN would be safer?

Reply:
I'm not sure that I agree. Only values generated by ACME are provided via command line arguments, nothing user-inputted.

Additionally, serializing it in JSON, incurs additional overhead in the Executable (i.e., installing `jq`).

Reply:
Personally I don't envision people writing these plugins in bash really.

I think we should provide a really simple template project, probably in Go to make things easy, that people can fork to create their own 'shim'.

Heptio have done a similar thing for ark: https://github.com/heptio/ark-plugin-example